eclass/ecm-common.eclass

```diff
@@ -0,0 +1,347 @@
+	_EOF_
+
+	if [[ ${ECM_HANDBOOK_DIRS} ]]; then
+		for i in "${ECM_HANDBOOK_DIRS[@]}"; do
+			if [[ -d ${i} ]]; then
+				cat >> CMakeLists.txt <<- _EOF_ || die
+					add_subdirectory(${i})
+				_EOF_
+			fi
+		done
+	fi
+}
+
+# @FUNCTION: _ecm-common_generate_desktop_file
+# @INTERNAL
+# @DESCRIPTION:
+# Find KF[56]KCMUtils and iterate through ECM_KCM_TARGETS to generate
+# desktop files out of json.
+_ecm-common_generate_desktop_file() {
+	[[ ${ECM_KCM_TARGETS} ]] || return
+	local i
+
+	cat >> CMakeLists.txt <<- _EOF_ || die
+		find_package(KF\${KFSLOT}KCMUtils REQUIRED)
+		# extracted from kcmutils_generate_desktop_file(kcm_target)
+		function(ecmcommon_generate_desktop_file kcm_target subdir)
+			set(IN_FILE \${CMAKE_CURRENT_SOURCE_DIR}/\${subdir}\${kcm_target}.json)
+			set(OUT_FILE \${CMAKE_CURRENT_BINARY_DIR}/\${kcm_target}.desktop)
+			add_custom_target(\${kcm_target}-kcm-desktop-gen ALL
+				COMMAND KF\${KFSLOT}::kcmdesktopfilegenerator \${IN_FILE} \${OUT_FILE}
+				DEPENDS \${IN_FILE})
+			install(FILES \${OUT_FILE} DESTINATION \${KDE_INSTALL_APPDIR})
+		endfunction()
+	_EOF_
+
+	for i in "${ECM_KCM_TARGETS[@]}"; do
+		cat >> CMakeLists.txt <<- _EOF_ || die
+			ecmcommon_generate_desktop_file(${i%:*} ${i#*:})
+		_EOF_
+	done
+}
+
+# @FUNCTION: _ecm-common_ecm_install_icons
+# @INTERNAL
+# @DESCRIPTION:
+# Installs icons listed in ECM_INSTALL_ICONS using ecm_install_icons
+_ecm-common_ecm_install_icons() {
+	[[ ${ECM_INSTALL_ICONS} ]] || return
+	local i
+	for i in "${ECM_INSTALL_ICONS[@]}"; do
+		cat >> CMakeLists.txt <<- _EOF_ || die
+			ecm_install_icons(ICONS ${i%:*} DESTINATION ${i#*:})
+		_EOF_
+	done
+}
+
+# @FUNCTION: _ecm-common_ecm_install_files
+# @INTERNAL
+# @DESCRIPTION:
+# Installs files listed in ECM_INSTALL_FILES using install(FILES ...)
+_ecm-common_ecm_install_files() {
+	[[ ${ECM_INSTALL_FILES} ]] || return
+	local i
+	for i in "${ECM_INSTALL_FILES[@]}"; do
+		cat >> CMakeLists.txt <<- _EOF_ || die
+			install(FILES ${i%:*} DESTINATION ${i#*:})
+		_EOF_
+	done
+}
+
+# @FUNCTION: ecm-common_inject_heredoc
+# @DESCRIPTION:
+# Override this to inject custom Heredoc into the root CMakeLists.txt
+ecm-common_inject_heredoc() {
+	debug-print-function ${FUNCNAME} "$@"
+}
+
+# @FUNCTION: _ecm-common_summary
+# @INTERNAL
+# @DESCRIPTION:
+# Just calls feature_summary
+_ecm-common_summary() {
+	cat >> CMakeLists.txt <<- _EOF_ || die
+
+		feature_summary(WHAT ALL INCLUDE_QUIET_PACKAGES FATAL_ON_MISSING_REQUIRED_PACKAGES)
+	_EOF_
+}
+
+# @FUNCTION: ecm-common_inject_heredoc
+# @DESCRIPTION:
+# Override this to add more KF6 has_version checks to pkg_setup(),
+# corresponding with any additional KF6_BDEPEND defined pre-inherit.
+# If false, we'll assume KF5 dependencies are fulfilled via BDEPEND.
+ecm-common-check_deps() {
+	return 1;
+}
+
+ecm-common_pkg_setup() {
+	if has_version -b "kde-frameworks/ki18n:6" &&
+		( in_iuse !handbook || ( in_iuse handbook && use handbook && has_version -b "kde-frameworks/kdoctools:6" ) ) &&
```

Comment:
Jeez!

Maybe split into nested functions (although they don't really exist in bash) within `ecm-common_pkg_setup`, so the condition reads nicer.

Reply:
Want to come up with a suggestion? I'm struggling there to put more beauty into it for a hopefully only shortlived time where we have KF5-based alternatives even available.

Reply:
Inside `ecm-common_pkg_setup`:
```
_handbook_kf6() {
	return in_iuse handbook && use handbook && has_version -b "kde-frameworks/kdoctools:6"
}
```

and so on, then use that in the `if`

Reply:
tidied up now with @ionenwks's suggestion, hope that's good enough